### PR TITLE
Add specs

### PR DESCRIPTION
**Pull request by VasylPosmit**
Add specs
- [x]  getDefaultController() specs still needed.

### CODE REVIEW

File: src/app/common/user.service.spec.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        userService.user.nutrition.weight = 40;
+        userService.user.nutrition.height = 150;
+        expect(userService.calculateBMI()).toEqual(17.8);
+      });
+      it('Undefined without any input', function() {
+        expect(userService.calculateBMI()).toBeUndefined();
+        userService.user.nutrition.height = 150;
+        expect(userService.calculateBMI()).toBeUndefined();
+        userService.user.nutrition.weight = 40;
+        expect(userService.calculateBMI()).toBeDefined();
```
Comment (andrenovax) [inline src/app/common/user.service.spec.js]: 4 different tests:
- user has height, does not have weight
- user has weight, does not have height
- user does not have weight and height
- user has weight and height and formula is calculated properly

File: src/app/common/user.service.spec.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      });
+    });
+
+    describe('calculate BMR', function() {
+      it('is function', function() {
+        expect(userService.calculateBMR).toEqual(jasmine.any(Function));
+      });
+      it('undefined without inputs', function() {
+        expect(userService.calculateBMR()).toBeUndefined();
+      });
```
Comment (andrenovax) [inline src/app/common/user.service.spec.js:78]: same as above

File: src/app/common/user.service.spec.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
+(function() {
```
Comment (andrenovax) [inline src/app/common/user.service.spec.js:1]: one unit test should test one case

File: src/app/common/user.service.spec.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      });
+    });
+    describe('sleepDuration', function() {
+      it('is a function', function() {
+        expect(userService.getUser).toEqual(jasmine.any(Function));
+      });
+      it('returns null without properties defined', function() {
+        expect(userService.sleepDuration()).toBeNull();
+        userService.sleepDuration();
+        expect(userService.sleepDuration()).toBeNull();
```
Comment (andrenovax) [inline src/app/common/user.service.spec.js:141]: what is going on here?

File: src/app/common/user.service.spec.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    });
+
+    describe('getUser()', function() {
+      it('is a function', function() {
+        expect(userService.getUser).toEqual(jasmine.any(Function));
+      });
+      it('has not been called', function() {
+        spyOn(userService, 'getUser');
+        expect(userService.getUser).not.toHaveBeenCalled();
+      });
```
Comment (andrenovax) [inline src/app/common/user.service.spec.js:175]: ???

File: src/app/common/user.service.spec.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      it('triggers calculateBMI and change user.BMI', function() {
+        expect(userService.user.BMI).toBeNull();
+        spyOn(userService, 'calculateBMI').and.callThrough();
+        userService.getUser();
+        expect(userService.calculateBMI).toHaveBeenCalled();
```
Comment (andrenovax) [inline src/app/common/user.service.spec.js]: you do not have to do that,
just check the values before and after the call

File: src/app/components/recommendations/recommendatons.directive.spec.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        state = 'nutrition';
+        scope.layout = {
+          user: userService.user,
+          data: sectionsService.data[state]
+        };
+        scope.$digest();
+      });
+    });
+
+    it('should applied template', function () {
```
Comment (andrenovax) [inline src/app/components/recommendations/recommendatons.directive.spec.js:24]: ?

File: src/app/components/recommendations/recommendatons.directive.spec.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      expect(directiveElem.html()).not.toEqual('');
+    });
+
+
+    it('should compile text', function () {
+      expect(directiveElem.find('h2').text()).toContain('Health recommendations for ');
```
Comment (andrenovax) [inline src/app/components/recommendations/recommendatons.directive.spec.js]: you should try to be independent from the markup and the hardcoded text,
all you want to know is:
- if the directive is rendered or not.
- if the directive displays all the values
- if the directive displays correct values for every case (ng-show, ng-if)

btw, try to use `ng-if`, it's better for performance reasons

Comment (andrenovax) [inline src/app/components/recommendations/recommendatons.directive.spec.js]: never test libraries functionality

File: src/app/components/sideNav/sidenav.services.spec.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      spyOn(sidenavService, 'toggleList');
+      expect(sidenavService.toggleList).not.toHaveBeenCalled();
+
+      spyOn(sidenavService, 'openLeftMenu');
+      expect(sidenavService.openLeftMenu).not.toHaveBeenCalled();
+
+      spyOn(sidenavService, 'closeSidenav');
+      expect(sidenavService.closeSidenav).not.toHaveBeenCalled();
+    });
+    it('toggleList should not trigger open/close functions', function() {
```
Comment (andrenovax) [inline src/app/components/sideNav/sidenav.services.spec.js:28]: nonsense

File: src/app/components/sideNav/sidenav.services.spec.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      spyOn(sidenavService, 'openLeftMenu');
+      spyOn(sidenavService, 'closeSidenav');
+
+      sidenavService.toggleList();
+      expect(sidenavService.openLeftMenu).not.toHaveBeenCalled();
+      expect(sidenavService.closeSidenav).not.toHaveBeenCalled();
+    });
```
Comment (andrenovax) [inline src/app/components/sideNav/sidenav.services.spec.js:35]: check if every function does what it has to do

File: src/app/components/sideNav/sidenav.services.spec.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  });
+})();
```
Comment (andrenovax) [inline src/app/components/sideNav/sidenav.services.spec.js:37]: btw, move "left" in variables on top in the service

File: src/app/defaultController.spec.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+      $provide.value('dataKey', dataKey);
+    }));
+
+
+    beforeEach( inject( function( ) {
+    }));
+
+
+
+    describe('values: ', function(){
```
Comment (andrenovax) [inline src/app/defaultController.spec.js:22]: xdescribe?

File: src/app/defaultController.spec.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+      });
+      xit('dataKey is a string', function() {
+        expect(controller.state).toEqual('default');
+      });
+      xit('self.state is a string', function () {
+        expect(controller.state).toBe(jasmine.any(String));
+      });
+      xit('getDefaultController has state', function(){
+        expect(controller.state).toEqual(jasmine.any(String));
+      });
```
Comment (andrenovax) [inline src/app/defaultController.spec.js:34]: ? ? ? ?
just give one state and test it

File: src/app/defaultController.spec.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+      xit('self.data should be object', function(){
+        expect(self.data).toEqual(jasmine.any(Object));
```
Comment (andrenovax) [inline src/app/defaultController.spec.js:36]: ???
check if you use the correct data, once!

File: src/app/defaultController.spec.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+      });
+      xit('self.calculate', function(){
+        expect(self.calculate).not.toHaveBeenCalled();
+      });
+      xit('assigns a person to the scope', function () {
+        expect(controller instanceof getDefaultController).toBeTruthy();
+      });
+    });
+
+    xdescribe('has activate', function() {
```
Comment (andrenovax) [inline src/app/defaultController.spec.js:52]: nonsense

File: src/app/defaultController.spec.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+      it('activate should not be undefined', function() {
+        expect(controller.activate).not.toBeUndefined();
+      });
+    });
+  });
```
Comment (andrenovax) [inline src/app/defaultController.spec.js:57]: where is the test for calculate function?

File: src/app/common/user.service.spec.js (hunk 1 above)
Comment (VasylPosmit) [inline src/app/common/user.service.spec.js]: - [x] done  https://github.com/VasylPosmit/health/pull/17/commits/727854b3748aa4c25848052c386468ca76011a08

File: src/app/common/user.service.spec.js (hunk 2 above)
Comment (VasylPosmit) [inline src/app/common/user.service.spec.js:78]: - [x] done
  https://github.com/VasylPosmit/health/pull/17/commits/cd9a76086d00036dcf37dbef80928e7ea1cfce28

File: src/app/common/user.service.spec.js (hunk 3 above)
Comment (VasylPosmit) [inline src/app/common/user.service.spec.js:1]: john papa tell the same thing

> Write a set of tests for every story. 
> [john papa](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#style-y190)

Specifically. Do I need  to exclude next afterEach?
`afterEach(function(){
       userService.user = user;
    });`

File: src/app/common/user.service.spec.js (hunk 5 above)
Comment (VasylPosmit) [inline src/app/common/user.service.spec.js:175]: It was the reminder.

Sorry, I do remember:

> Не держи код на прозапас (это мусор) (с) Novak

Please check new [specs for sleepDuration()](https://github.com/VasylPosmit/health/pull/17/commits/36e1b555e03ba7e0a283819d2836778ec871add0)

File: src/app/common/user.service.spec.js (hunk 4 above)
Comment (VasylPosmit) [inline src/app/common/user.service.spec.js:141]: - [x] rewrite

I would appreciate you leaving the action that you expect from me regarding the comment.
For example: 
- Explain 
- Rewrite code
- Refactor 
- Read john papa Testing section and rewrite specs names

File: src/app/common/user.service.spec.js (hunk 6 above)
Comment (VasylPosmit) [inline src/app/common/user.service.spec.js]: - [x] refactor

[Now it's much cleaner.](https://github.com/VasylPosmit/health/pull/17/commits/4a3de23aa1a951807d6161095e90c12fbdb59bd1)
KISS

File: src/app/components/recommendations/recommendatons.directive.spec.js (hunk 8 above)
Comment (VasylPosmit) [inline src/app/components/recommendations/recommendatons.directive.spec.js]: - [x] rewrite specs 
- [x] add specs

I beg you pardon, but where do I test library functionality?

File: src/app/common/user.service.spec.js (hunk 3 above)
Comment (andrenovax) [inline src/app/common/user.service.spec.js:1]: what Specifically ?
bro, if you test what happens if one argument is falsy, you should not test what happens if other variables are falsy in the same test, etc

File: src/app/components/sideNav/sidenav.controller.spec.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    describe('toggle', function(){
+      it('sidenavController do not activate toggleSidenav()', function() {
+        expect(controller.toggleSidenav).not.toHaveBeenCalled();
+      });
+
+      it('toggleSidenav trigger sidenavService', function() {
+        controller.toggleSidenav();
+        expect(controller.toggleSidenav).toHaveBeenCalled();
+      });
+    });
```
Comment (andrenovax) [inline src/app/components/sideNav/sidenav.controller.spec.js:31]: ???

File: src/app/components/sideNav/sidenav.services.spec.js (hunk 10 above)
Comment (VasylPosmit) [inline src/app/components/sideNav/sidenav.services.spec.js:35]: Spend 2 h figuring out why this test doesn't pass.

```
//sidenav.service.js
self.toggleList   = toggleSidenav;
 function toggleSidenav() {
      console.log($mdSidenav('left').isOpen()); //  console: false
      $mdSidenav(componentId).toggle();
      console.log($mdSidenav('left').isOpen()); //  console: true
      $
    }
```

[$mdSidenav API](https://material.angularjs.org/latest/api/service/$mdSidenav)

```
//sidenav.service.specs.js
    beforeEach( function(){
      module( 'healthGuide' );
      inject( function (_sidenavService_, _$mdSidenav_) {
        sidenavService = _sidenavService_;
        mdSidenav = _$mdSidenav_('left');                // works
      });
    });
    it('openLeftMenu should open sideNav', function() {
       expect(mdSidenav.isOpen()).toBeNull();
       sidenavService.toggleList();
       expect(mdSidenav.isOpen()).toBeTruthy();          // FAIL. Expected null to be truthy.
    });

```

My conclusion that jasmine access two different instances of $mdSidenav.

File: src/app/components/sideNav/sidenav.services.spec.js (hunk 11 above)
Comment (VasylPosmit) [inline src/app/components/sideNav/sidenav.services.spec.js:37]: - [x] done

File: src/app/defaultController.spec.js (hunk 12 above)
Comment (VasylPosmit) [inline src/app/defaultController.spec.js:22]: I still can not access getDefaultController(state) for testing.(8h+++ spend on this issue)

File: src/app/components/sideNav/sidenav.services.spec.js (hunk 10 above)
Comment (andrenovax) [inline src/app/components/sideNav/sidenav.services.spec.js:35]: dude, https://github.com/andrenovax/dopomoga_project/blob/master/client/src/app/components/navigation/navigation.spec.js

Comment (VasylPosmit) [inline src/app/components/sideNav/sidenav.services.spec.js:35]: omg
- [x] done https://github.com/VasylPosmit/health/pull/21

**Pull request merged**